### PR TITLE
Distribute bootstrap in CI

### PR DESCRIPTION
**Pull request by dvtkrlbs**
This pre-compiles bootstrap from source and adds it to the existing `rust-dev` component. There are two main goals here:
1. Make it faster to build rust from source, both the first time and incrementally
2. Make it easier to add non-python entrypoints, since they can call out to bootstrap directly rather than having to figure out the right flags to pre-compile it. This second part is still in a bit of flux, see the tracking issue below for more information.

There are also several changes to make bootstrap able to run on a machine other than the one it was built (particularly around `config.src` and `config.out` detection). I (@jyn514) am slightly concerned these will regress unless tested - maybe we should add an automated test that runs bootstrap in a chroot or something? Unclear whether the effort is worth the test coverage.

Helps with https://github.com/rust-lang/rust/issues/94829.

### CODE REVIEW

File: src/bootstrap/dist.rs [hunk 1]
```diff
@@ -2009,17 +2009,30 @@ impl Step for RustDev {
...
+
+        tarball.set_overlay(OverlayKind::LLVM);
+
         /* run only if llvm-config isn't used */
         if let Some(config) = builder.config.target_config.get(&target) {
             if let Some(ref _s) = config.llvm_config {
-                builder.info(&format!("Skipping RustDev ({}): external LLVM", target));
-                return None;
+                builder.info(&format!("Skipping llvm ({}): external LLVM", target));
+                return Some(tarball.generate());
```
Comment (jyn514) [inline src/bootstrap/dist.rs:2032]: I think this isn't the right behavior. With your changes now, we could end up with a tarball that has just bootstrap and not LLVM. In CI that doesn't matter, because we always build LLVM from source, but if a distro is running `x.py dist` from source, they'll end up with the rust-dev component even though they don't actually need it.

I think the proper fix is to add a `default_condition` to `should_run` above that doesn't run this step by default when using external LLVM. That will still let you test this locally while building LLVM, and won't break CI, but also avoids building rust-dev without LLVM unless you specifically ask for it.

File: src/bootstrap/dist.rs [hunk 2]
```diff
@@ -2009,17 +2009,30 @@ impl Step for RustDev {
     fn run(self, builder: &Builder<'_>) -> Option<GeneratedTarball> {
         let target = self.target;
 
+        let mut tarball = Tarball::new(builder, "rust-dev", &target.triple);
+
+        let bootstrap_outdir = &builder.bootstrap_out;
+        for file in &[
+            "bootstrap",
+            "libbootstrap.rlib",
```
Comment (jyn514) [inline src/bootstrap/dist.rs:2017]: The rlib shouldn't be necessary, bootstrap is statically linked.
```suggestion
```

**Review by jyn514 — COMMENTED** [review]
Thanks! This looks like roughly the right approach - I left a couple comments, but I'll go ahead and do a try run to confirm this works in CI.

File: src/bootstrap/dist.rs (hunk 1 above)
Comment (dvtkrlbs) [reply src/bootstrap/dist.rs:2031]: hmm how i can access to the `target_config` from `should_run`?

Comment (dvtkrlbs) [reply src/bootstrap/dist.rs:2031]: nvm ShouldRun has a builder field

Comment (jyn514) [reply src/bootstrap/dist.rs]: We chatted a bit more about this [on Zulip](https://rust-lang.zulipchat.com/#narrow/stream/122652-new-members/topic/distribute.20bootstrap.20in.20CI.20.2398483/near/287448035). This is non-trivial to get working because to set a `default_condition` we need to know which platform we're targeting. But bootstrap is target-independent (wants to be `ONLY_HOSTS = true`) while LLVM is target-dependent; plus if we turn this off by default, I'd worry that we'd miss adding it in one of the builders.

@Mark-Simulacrum how do you feel about separating the LLVM component from the bootstrap component? I know the original motivation for calling this `rust-dev` was for bundling them, but that seems like more effort than it's worth ... plus it means you have to download and unpack LLVM even if you're only planning to use bootstrap and compile llvm from source. I think having a separate `llvm-nightly` and `bootstrap-nightly` component makes more sense, and is easier to maintain.

**Comment by jyn514** [comment]
This is looking pretty good to me except for the defaults. @Mark-Simulacrum I'd appreciate if you could take a look at https://github.com/rust-lang/rust/pull/98483#discussion_r906853614.

File: src/bootstrap/config.rs [hunk 3]
```diff
@@ -798,7 +798,20 @@ impl Config {
...
+            .unwrap_or_else(|_| {
+                let base_path = output(Command::new("git").arg("rev-parse").arg("--show-toplevel"))
+                    .trim()
+                    .to_owned();
+
+                config.src = PathBuf::from(base_path);
+                config.initial_rustc =
+                    config.src.join("build").join(config.build.triple).join("stage0/bin/rustc");
+                config.initial_cargo =
+                    config.src.join("build").join(config.build.triple).join("stage0/bin/cargo");
```
Comment (jyn514) [inline src/bootstrap/config.rs:811]: this should be `config.out` instead, the build directory isn't always within the source checkout (particularly in CI).
```suggestion
                    config.out.join(config.build.triple).join("stage0/bin/rustc");
                config.initial_cargo =
                    config.out.join(config.build.triple).join("stage0/bin/cargo");
```

**Review by jyn514 — CHANGES_REQUESTED** [review]

File: src/bootstrap/dist.rs (hunk 1 above)
Comment (Mark-Simulacrum) [reply src/bootstrap/dist.rs]: I'm OK with separating them; I don't think I would suffix with -nightly. Maybe -internal could work well; the indicator we want to give is "not really intended for consumption outside of bootstrap itself".

File: src/bootstrap/config.rs [hunk 4]
```diff
@@ -798,7 +798,20 @@ impl Config {
...
+                    .trim()
+                    .to_owned();
+
+                config.src = PathBuf::from(base_path);
+                config.initial_rustc =
+                    config.out.join(config.build.triple).join("stage0/bin/rustc");
+                config.initial_cargo =
+                    config.out.join(config.build.triple).join("stage0/bin/cargo");
+                t!(std::fs::read(&config.src.join("src").join("stage0.json")))
+            });
```
Comment (Mark-Simulacrum) [inline src/bootstrap/config.rs:813]: This looks pretty suspicious to me -- if we need to remap the base path via git rev-parse, that's probably something that should be done when discovering config.src initially, not when failing to read stage0.json.

We should also most likely hook the discovery mechanism to not be opportunistic (on failing to find stage0.json), but rather based on whether we're building for distribution (i.e., in CI), in which case we can be reliant on git and not try to to use the env-variables set by Cargo for these things.

File: src/bootstrap/dist.rs (hunk 1 above)
Comment (Mark-Simulacrum) [reply src/bootstrap/dist.rs]: That said, I'm not sure that we need to? It seems to me that it'd be totally fine to only produce this component when enabled (perhaps a new config.toml option for ease of passing it to all the dist builders).

**Comment by Mark-Simulacrum** [comment]
It'd be great to flesh out the PR description a little with details of what this is trying to aim for -- especially in bootstrap, details there really do help when looking back on PRs that seem innocuous at the time of writing a few years down the line.

File: src/bootstrap/dist.rs (hunk 1 above)
Comment (jyn514) [reply src/bootstrap/dist.rs]: @Mark-Simulacrum so, my original concern was that someone would run `x dist` intending to get a toolchain and get `rust-dev` packaged along with it. Right now that doesn't usually happen because most distros are using a system LLVM instead of building from source. After this change, they'll get rust-dev even with system llvm (it will have just bootstrap, not the pre-compiled LLVM).

If you think that's ok, we don't need to change anything and can keep this as a default step.

File: src/bootstrap/config.rs (hunk 4 above)
Comment (jyn514) [reply src/bootstrap/config.rs:813]: > We should also most likely hook the discovery mechanism to not be opportunistic (on failing to find stage0.json), but rather based on whether we're building for distribution (i.e., in CI), in which case we can be reliant on git and not try to to use the env-variables set by Cargo for these things.

@bjorn3 objected to that on zulip:
> Both options are incorrect for CI distributed rustbuild when you are invoking it from another directory than the source directory.

but I agree just using git is more reliable and simpler.

File: src/bootstrap/dist.rs (hunk 1 above)
Comment (Mark-Simulacrum) [reply src/bootstrap/dist.rs]: That's also fine; I'm not really worried about us producing a few extra artifacts. If we want to avoid it, I think the right way is to add a config.toml option (internal-artifacts or whatever name) to gate this step.

File: src/bootstrap/config.rs (hunk 4 above)
Comment (Mark-Simulacrum) [reply src/bootstrap/config.rs:813]: I think the future option is probably that we embed a symlink (or, well, text file, hard link, whatever) next to bootstrap pointing to the real source (generated by whatever tool is downloading bootstrap), but for now I'm OK sticking to just using git for this since we don't actually have that tool.

But I do think this should be refactored to be in the right place.

File: src/bootstrap/config.rs [hunk 5]
```diff
@@ -765,13 +765,14 @@ impl Config {
 
         // set by build.rs
         config.build = TargetSelection::from_user(&env!("BUILD_TRIPLE"));
-        let manifest_dir = PathBuf::from(env!("CARGO_MANIFEST_DIR"));
+        let base_path =
+            output(Command::new("git").arg("rev-parse").arg("--show-toplevel")).trim().to_owned();
         // Undo `src/bootstrap`
```
Comment (jyn514) [inline src/bootstrap/config.rs:770]: this comment is out of date
```suggestion
```

File: src/bootstrap/config.rs [hunk 6]
```diff
@@ -765,13 +765,14 @@ impl Config {
...
-        config.src = manifest_dir.parent().unwrap().parent().unwrap().to_owned();
+        config.src = PathBuf::from(base_path);
         config.out = PathBuf::from("build");
 
-        config.initial_cargo = PathBuf::from(env!("CARGO"));
-        config.initial_rustc = PathBuf::from(env!("RUSTC"));
+        config.initial_rustc = config.out.join(config.build.triple).join("stage0/bin/rustc");
+        config.initial_cargo = config.out.join(config.build.triple).join("stage0/bin/cargo");
```
Comment (jyn514) [inline src/bootstrap/config.rs:775]: This is still in `default-opts`, so `config.out` is always `build`. It needs to be set later, after `config.out` is set to the runtime value: https://github.com/rust-lang/rust/blob/db896fd318b1fea977a65d01d030b827d2a5b03f/src/bootstrap/config.rs#L862-L868

File: src/bootstrap/config.rs [hunk 7]
```diff
@@ -765,13 +765,14 @@ impl Config {
 
         // set by build.rs
         config.build = TargetSelection::from_user(&env!("BUILD_TRIPLE"));
-        let manifest_dir = PathBuf::from(env!("CARGO_MANIFEST_DIR"));
+        let base_path =
+            output(Command::new("git").arg("rev-parse").arg("--show-toplevel")).trim().to_owned();
```
Comment (jyn514) [inline src/bootstrap/config.rs:769]: I think this will break vendoring. We should fallback to `CARGO_MANIFEST_DIR` if git exits with an error.

**Review by jyn514 — DISMISSED** [review]

**Comment by dvtkrlbs** [comment]
i will cleanup the cargo manifest fallback part i just wanted to test if the CI passes

**Comment by jyn514** [comment]
@dvtkrlbs are you planning to keep working on this? I would love to get this merged so we can start moving away from the python entrypoint.

**Comment by dvtkrlbs** [comment]
sorry i got stuck on a part of it and then forgot about it. I can pick it up again. My issu was it using wrong rustc or something like that.

File: src/bootstrap/config.rs (hunk 7 above)
Comment (jyn514) [reply src/bootstrap/config.rs:769]: you marked these comments as resolved, but I don't see any code changes?

**Comment by jyn514** [comment]
try seems to have gotten stuck?

@bors try

**Comment by jyn514** [comment]
oh, it finished but just didn't leave a comment. the `rust-dev` component is missing because it's not longer a default step - the `dist` steps all need to be updated to use `dist rust-dev --include-default-paths` instead.

**Comment by jyn514** [comment]
@dvtkrlbs I went ahead and fixed the test failure. The issue was that we were unconditionally setting `initial_rustc` to a path relative to the build directory, but we should have only been setting it when we were in a worktree other than the one bootstrap was compiled in.

Next steps are for me to run `@bors try` and make sure this works locally for me with a binary that was compiled on a different machine. But I already tested different working directories locally so I'm not too worried.

**Comment by dvtkrlbs** [comment]
Ah cool sorry for not looking at it muself. Was really busy with work.

**Comment by jyn514** [comment]
No worries! I appreciate all the work you've already done :)

**Comment by jyn514** [comment]
This still isn't working because we don't properly detect the source directory:
```
; strace -e openat build/bootstrap/bin/bootstrap c 2>&1 | tail -n4 | head -n2
openat(AT_FDCWD, "/checkout/src/stage0.json", O_RDONLY|O_CLOEXEC) = -1 ENOENT (No such file or directory)
thread 'main' panicked at 'std::fs::read(&config.src.join("src").join("stage0.json")) failed with No such file or directory (os error 2)', config.rs:835:27
```
The problem is we're trying to run git in a directory set at compile time: https://github.com/rust-lang/rust/blob/f21fcc0e1fc55f3588972f174b3765827260274f/src/bootstrap/config.rs#L777
but when running on a different machine, that directory won't exist. I think we need to keep using the current working directory, and just not support running from outside the git repo for downloadable bootstrap.

The only edge case I can imagine that breaking is if someone runs bootstrap from a different checkout of rust than their current working directory, but that seems ... weird .... anyway, and I don't think we should bother supporting it.

**Comment by jyn514** [comment]
Hmm, so another thing I noticed just now - because LLVM and bootstrap are in the same rust-dev component, we have to rebuild LLVM whenever bootstrap changes, which makes CI a lot slower.

@pietroalbini @Mark-Simulacrum I know that you originally imagined `rust-dev` as a single component for all the work people would have to do on bootstrap, but between this and the step defaults it just seems more confusing than before ... do you mind if I separate this into its own `bootstrap` dist component?

**Comment by Mark-Simulacrum** [comment]
Yeah, splitting them is a good idea. We should probably rename rust-dev at some point (will need to be careful to avoid breaking people).

**Comment by jyn514** [comment]
Ok, since https://github.com/rust-lang/rust/pull/98483/commits/0d3bd7c46733b468066df4d348f4e40ede342752 this is working great for me locally :) I tested with
```
rustup-toolchain-install-master -c rust-dev -- 07eb275f0c8067d790ecabd640a0831603d79ec8
cp -r $(rust +07eb275f0c8067d790ecabd640a0831603d79ec8 --print sysroot)/bootstrap/ build/
ln -s build/bootstrap/bin build/bootstrap/debug  # This step is unnecessary after https://github.com/rust-lang/rust/pull/98483/commits/0d3bd7c46733b468066df4d348f4e40ede342752
build/bootstrap/bin/bootstrap check
```

The next steps are to separate the rust-dev and bootstrap components (which will end up undoing some of your work @dvtkrlbs, sorry), and then I think this is ready to merge!

**Comment by dvtkrlbs** [comment]
Oh so differrent step that does not include LLVM at all right ?

**Comment by jyn514** [comment]
@dvtkrlbs yes, exactly - the LLVM step should be the same as it was originally, and the new `bootstrap` component will use the same logic you used for `rust-dev`: off by default, but built explicitly by each builder.

File: src/bootstrap/dist.rs [hunk 8]
```diff
@@ -2023,17 +2023,14 @@ impl Step for RustDev {
...
+
+        let bootstrap_outdir = &builder.bootstrap_out;
+        for file in &["bootstrap", "llvm-config-wrapper", "rustc", "rustdoc", "sccache-plus-cl"] {
+            tarball.add_file(bootstrap_outdir.join(file), "bootstrap/bin", 0o755);
         }
 
-        let mut tarball = Tarball::new(builder, "rust-dev", &target.triple);
         tarball.set_overlay(OverlayKind::LLVM);
-
         builder.ensure(crate::native::Llvm { target });
```
Comment (jyn514) [inline src/bootstrap/dist.rs:2034]: now that you've added the new `Bootstrap` step, you can undo all changes to `rust-dev` and have it stay as LLVM only. You'll also need to update all the CI changes to use `bootstrap` instead of `rust-dev`. Sorry this keeps changing :( I think this should hopefully be the last time.

File: src/bootstrap/config.rs [hunk 9]
```diff
@@ -770,21 +770,22 @@ impl Config {
...
+        config.initial_rustc = PathBuf::from(env!("RUSTC"));
+
         let manifest_dir = PathBuf::from(env!("CARGO_MANIFEST_DIR"));
         // Undo `src/bootstrap`
         config.src = manifest_dir.parent().unwrap().parent().unwrap().to_owned();
         config.out = PathBuf::from("build");
 
-        config.initial_cargo = PathBuf::from(env!("CARGO"));
-        config.initial_rustc = PathBuf::from(env!("RUSTC"));
-
```
Comment (jyn514) [inline src/bootstrap/config.rs:780]: nit: you can undo this change

**Review by jyn514 — CHANGES_REQUESTED** [review]

File: src/ci/docker/host-x86_64/dist-x86_64-linux/Dockerfile [hunk 10]
```diff
@@ -82,7 +82,7 @@ ENV RUST_CONFIGURE_ARGS \
 ENV SCRIPT ../src/ci/pgo.sh python3 ../x.py dist \
     --host $HOSTS --target $HOSTS \
     --include-default-paths \
-    build-manifest
+    build-manifest rust-dev
```
Comment (jyn514) [inline src/ci/docker/host-x86_64/dist-x86_64-linux/Dockerfile:85]:
```suggestion
    build-manifest bootstrap
```

File: src/ci/github-actions/ci.yml [hunk 11]
```diff
@@ -675,8 +675,9 @@ jobs:
                 --host=aarch64-pc-windows-msvc
                 --enable-full-tools
                 --enable-profiler
-              SCRIPT: python x.py dist
-              DIST_REQUIRE_ALL_TOOLS: 1
+              SCRIPT: python x.py dist bootstrap --include-default-paths
+              # RLS does not build for aarch64-pc-windows-msvc. See rust-lang/rls#1693
+              DIST_REQUIRE_ALL_TOOLS: 0
```
Comment (jyn514) [inline src/ci/github-actions/ci.yml:680]: this looks unintentional

File: src/bootstrap/dist.rs (hunk 8 above)
Comment (jyn514) [reply src/bootstrap/dist.rs:1999]: I still see changes here :(

**Review by jyn514 — DISMISSED** [review]
LGTM with these last few comments fixed and the commits squashed :) thanks again for sticking with this!

File: src/ci/github-actions/ci.yml (hunk 11 above)
Comment (dvtkrlbs) [reply src/ci/github-actions/ci.yml:680]: do i need to remove bootstrap from this workflow step ?

File: src/ci/github-actions/ci.yml [hunk 12]
```diff
@@ -676,6 +676,7 @@ jobs:
                 --enable-full-tools
                 --enable-profiler
               SCRIPT: python x.py dist
+              # RLS does not build for aarch64-pc-windows-msvc. See rust-lang/rls#1693
               DIST_REQUIRE_ALL_TOOLS: 1
```
Comment (jyn514) [inline src/ci/github-actions/ci.yml:680]: I meant this:
```yaml
              SCRIPT: python x.py dist bootstrap --include-default-paths
              DIST_REQUIRE_ALL_TOOLS: 1
```

File: src/bootstrap/dist.rs [hunk 13]
```diff
@@ -2050,6 +2050,41 @@ impl Step for RustDev {
...
+    pub target: TargetSelection,
+}
+
+impl Step for Bootstrap {
+    type Output = Option<GeneratedTarball>;
+    const DEFAULT: bool = false;
+    const ONLY_HOSTS: bool = true;
+
+    fn should_run(run: ShouldRun<'_>) -> ShouldRun<'_> {
+        run.alias("boostrap")
```
Comment (jyn514) [inline src/bootstrap/dist.rs:2067]:
```suggestion
        run.alias("bootstrap")
```

File: src/bootstrap/dist.rs [hunk 14]
```diff
@@ -2050,6 +2050,41 @@ impl Step for RustDev {
...
+    }
+
+    fn make_run(run: RunConfig<'_>) {
+        run.builder.ensure(Bootstrap { target: run.target });
+    }
+
+    fn run(self, builder: &Builder<'_>) -> Option<GeneratedTarball> {
+        let target = self.target;
+
+        let tarball = Tarball::new(builder, "boostrap", &target.triple);
```
Comment (jyn514) [inline src/bootstrap/dist.rs:2077]:
```suggestion
        let tarball = Tarball::new(builder, "bootstrap", &target.triple);
```

**Comment by jyn514** [comment]
Found the bug, the pre-existing `rustc` check forgot that executables have `.exe` on windows :facepalm: Fixed that and also fixed up the history a bit.

@bors r+

**Comment by jyn514** [comment]
```
-                                             // + span: $SRC_DIR/core/src/ptr/mod.rs:LL:COL
+                                             // + span: $LIB_DIRrary/core/src/ptr/mod.rs:487:1: 487:56
```
Looks like this broke our handling of `CFG_VIRTUAL_RUST_SOURCE_BASE_DIR` somehow - don't currently have time to look into it, but I'd suggest running compiletest before this change and after to see how the value of `CFG_VIRTUAL_RUST_SOURCE_BASE_DIR` changes.

File: src/bootstrap/config.rs [hunk 15]
```diff
@@ -805,7 +804,41 @@ impl Config {
...
+        cmd.arg("rev-parse").arg("--show-toplevel");
+        // Discard stderr because we expect this to fail when building from a tarball.
+        let output = cmd
+            .stderr(std::process::Stdio::null())
+            .output()
+            .ok()
+            .and_then(|output| if output.status.success() { Some(output) } else { None });
+        if let Some(output) = output {
+            let git_root = String::from_utf8(output.stdout).unwrap();
+            config.src = PathBuf::from(git_root.trim().to_owned())
```
Comment (mejrs) [inline src/bootstrap/config.rs:824]: I've tried running this; it seems to replace `"C:\\Users\\bruno\\Rust\\rustpr\\rust"` with `"C:/Users/bruno/Rust/rustpr/rust"`, which then causes a bunch of failing mir-opt tests.

Comment (jyn514) [reply src/bootstrap/config.rs:824]: We worked through this together and found that `crate::util::absolute(std::env::current_exe().unwrap().join("../../../.."))` does the same thing and also works on Windows, as well as supporting out-of-tree builds when building from a tarball :)

Comment (jyn514) [reply src/bootstrap/config.rs:824]: err no it doesn't actually do that last thing, the build directory need not be inside the source directory :( so we're going to hack this by transforming `\` to `/ when setting `CFG_VIRTUAL_RUST_SOURCE_BASE_DIR`.

Comment (mejrs) [reply src/bootstrap/config.rs:824]: I think the hacky solution for now is to just trim this manually. e.g. 
```suggestion
            let git_root = PathBuf::from(git_root.trim()).canonicalize().unwrap();
            let git_root = git_root.to_str().unwrap();

            // Bootstrap is quite bad at handling /? in front of paths
            config.src = match git_root.strip_prefix("\\\\?\\") {
                Some(p) => PathBuf::from(p),
                None => PathBuf::from(git_root),
            };
```

Comment (jyn514) [reply src/bootstrap/config.rs:824]: god I hate this lol

yeah that seems reasonable as a stop-gap

Comment (dvtkrlbs) [reply src/bootstrap/config.rs]: Committed. Sorry wasnt able to tes it myself lets see if ci passes.

File: src/bootstrap/config.rs [hunk 16]
```diff
@@ -821,7 +821,14 @@ impl Config {
             .and_then(|output| if output.status.success() { Some(output) } else { None });
         if let Some(output) = output {
             let git_root = String::from_utf8(output.stdout).unwrap();
-            config.src = PathBuf::from(git_root.trim().to_owned())
+            let git_root = PathBuf::from(git_root.trim()).canonicalize().unwrap();
```
Comment (jyn514) [inline src/bootstrap/config.rs:824]:
```suggestion
            // We need to canonicalize this path to make sure it uses backslashes instead of forward slashes.
            let git_root = PathBuf::from(git_root.trim()).canonicalize().unwrap();
```

**Comment by kleisauke** [comment]
This PR seems to break my workflow that configures the Rust nightly tarball inside a Git-tracked directory. For example:

```bash
$ git init my-project
$ mkdir my-project/rust-nightly
$ curl -sL https://static.rust-lang.org/dist/2022-09-28/rustc-nightly-src.tar.xz | tar xJC my-project/rust-nightly --strip-components=1
$ cd my-project/rust-nightly
$ git rev-parse --show-toplevel
/home/kleisauke/my-project
$ cp config.toml.example config.toml
$ ./x.py install --stage 1 -v
running: /home/kleisauke/my-project/rust-nightly/build/bootstrap/debug/bootstrap install --stage 1 -v
thread 'main' panicked at 'std::fs::read(&config.src.join("src").join("stage0.json")) failed with No such file or directory (os error 2)', config.rs:848:27
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
Traceback (most recent call last):
  File "./x.py", line 28, in <module>
    bootstrap.main()
  File "/home/kleisauke/my-project/rust-nightly/src/bootstrap/bootstrap.py", line 936, in main
    bootstrap(help_triggered)
  File "/home/kleisauke/my-project/rust-nightly/src/bootstrap/bootstrap.py", line 922, in bootstrap
    run(args, env=env, verbose=build.verbose, is_bootstrap=True)
  File "/home/kleisauke/my-project/rust-nightly/src/bootstrap/bootstrap.py", line 166, in run
    raise RuntimeError(err)
RuntimeError: failed to run: /home/kleisauke/my-project/rust-nightly/build/bootstrap/debug/bootstrap install --stage 1 -v
```
(this is a simplified example from our [Rust build script](https://github.com/libvips/build-win64-mxe/blob/master/build/plugins/llvm-mingw/rust.mk) based on [MXE](https://github.com/mxe/mxe))

Is there a way to skip this logic?

**Comment by jyn514** [comment]
@kleisauke we can fix your issue by using `current_dir(std::env::current_exe().parent()`. But in general running bootstrap outside the source directory has very limited support, bugs are fixed as they arise but we don't test it in CI.

**Comment by kleisauke** [comment]
@jyn514 IIUC, that wouldn't fix the above issue (+ I don't run the build script outside the source directory). The thing is that my parent directory is a Git-tracked directory and `git rev-parse --show-toplevel` would resolve to that, causing a build error since ` my-project/src/stage0.json` doesn't exist in that case.

Perhaps `git_root` needs to be checked for the presence of that file before setting `config.src`?

**Comment by jyn514** [comment]
Oh I see, you've unpacked a source tarball in a subdirectory of another git project. I highly suspect there's several other things that will be broken for you with that setup (like the commit hash we detect for the compiler version), it's not something we've ever considered or tested.

I'm willing to take a PR that falls back to the tarball behavior when `stage0.json` is missing, but this is very likely not the only thing that's broken and I'd suggest unpacking the tarball in a temporary directory instead.

**Comment by kleisauke** [comment]
Sounds reasonable. Fixed with commit https://github.com/libvips/build-win64-mxe/commit/9b293bca5a8c2889f3f3a846d698eec9c549f802 instead, thanks!

**Comment by cr1901** [comment]
>it's not something we've ever considered or tested.

I appear to be in the same boat as @kleisauke. I build rust in a directory called `build-rust`, which happens to be an (ignored) subdirectory of a git repo where I keep my scripts for building projects using LLVM. _The Rust source can be anywhere on the filesystem and is not a subdirectory of any other git repo in my case._

However, right now, my scripts for building Rust/LLVM only support building in the immediate subdirectory of where my scripts live. This commit makes it impossible for me to build Rust using my workflow unless I do `mv .git .git-bak` for my scripts repo every time I build Rust. Alternatively, I could just not version-control my scripts, but I don't want to do that :P.

**Comment by jyn514** [comment]
@cr1901 

> I'm willing to take a PR that falls back to the tarball behavior when stage0.json is missing, but this is very likely not the only thing that's broken and I'd suggest unpacking the tarball in a temporary directory instead.

**Comment by cr1901** [comment]
>this is very likely not the only thing that's broken

FWIW, doing `mv .git .git-bak` got me a successful compile. I've been using a separate build dir for several years to build Rust. So I'm willing to bet fallback logic will work at least for now.

>I'd suggest unpacking the tarball in a temporary directory instead.

If by "tarball" you mean "the Rust source" and not "the downloaded bootstrap compiler", I don't think this applies to me since I'm building Rust from the `git` source.

In other words, I think a PR that falls back to the tarball behavior is the only option I have unless you can think of a method to support my use case w/ a downloaded bootstrap.

That said, I use `sccache` for building Rust, so I don't know how much time I will lose having to rebuild the bootstrap artifacts when the bootstrap compiler version hasn't changed.

I'll make a PR when I get the chance.

**Comment by jyn514** [comment]
Ok, you have a slightly different case: the *build* directory is within an unrelated git repo, not the source directory. I agree we should support that.
